File: Easy/242-ValidAnagram.cc
#include <iostream>
#include <vector>
#include <string>
#include <unordered_map>

using namespace std;


// Description:  https://leetcode.com/problems/valid-anagram/description/



// Approach 1  

 /* 
    Time complexity:    O(n)
    Space complexity:   O(1)

    The strategy used to solve the problem involves counting the frequency of each character in both strings. 
    Two vectors of size 26 (for each letter in the alphabet) are used to store these frequencies. 
    The characters in the strings are iterated over, and their corresponding counts are incremented. 
    Finally, the two vectors are compared; if they are identical, the strings are anagrams
*/

class Solution {
public:
    bool isAnagram(string s, string t) {
        if (s.size() != t.size()) return false;

        vector<int> countS(26,0);
        vector<int> countT(26,0);

        for (int i = 0; i < s.size(); ++i) {
            countS[s[i] -'a']++;
            countT[t[i] -'a']++;
        }
        
        return countS == countT;
    }
};

// Approach 2

 /* 
    Time complexity:    O(n)
    Space complexity:   O(k) : where 'k' is the number of distinct characters across both input 

    This code replaces the fixed-size vectors with unordered_map to count the frequency of each 
    character in the input strings. This approach is more flexible and can handle any set of characters, 
    making it suitable for Unicode characters as well. 

*/

class Solution2 {
public:
    bool isAnagram(string s, string t) {
        if (s.size() != t.size()) return false;

        unordered_map<char, int> countS;
        unordered_map<char, int> countT;

        for (int i = 0; i < s.size(); ++i) {
            countS[s[i]]++;
            countT[t[i]]++;
        }

        return countS == countT;
    }
};




// Test
int main() {
    //Approach 1
    Solution solution;
    cout << boolalpha;
    cout << "Example 1: " << solution.isAnagram("anagram", "nagaram") << endl; // Output: true
    cout << "Example 2: " << solution.isAnagram("rat", "car") << endl; // Output: false

    //Approach 2
    Solution2 solution_unicode;
    cout << "Unicode Example 1: " << solution_unicode.isAnagram("मनीमनी", "नीम") << endl;       // Output: false
    cout << "Unicode Example 2: " << solution_unicode.isAnagram("心心相印", "心相心印") << endl; // Output: true
    cout << "Unicode Example 3: " << solution_unicode.isAnagram("🍏🍏🍏", "🍎🍎🍎") << endl;  // Output: false
    cout << "Unicode Example 4: " << solution_unicode.isAnagram("🍎🍏", "🍏🍎") << endl;      // Output: true
    
    return 0;
}

File: Java/Easy/242-ValidAnagram.java
// Description: https://leetcode.com/problems/valid-anagram/description/

/*
    Approach: Frequency Array

    First check if strings have equal length; if not, they can't be anagrams.
    Use two integer arrays of size 26 to count character frequencies.
    Iterate through both strings simultaneously, incrementing respective counts.
    Use character arithmetic (char - 'a') to map characters to array indices.
    Compare the frequency arrays using Arrays.equals() at the end.
    If identical, strings are anagrams; otherwise, they are not.

    Time complexity: O(n) - Single pass through both strings.
    Space complexity: O(1) - Fixed size arrays of 26 elements.
*/

import java.util.Arrays;

// Solution
class Solution {
    public boolean isAnagram(String s, String t) {
        if (s.length() != t.length()) return false;

        int[] countS = new int[26];
        int[] countT = new int[26];

        for (int i = 0; i < s.length(); i++) {
            countS[s.charAt(i) - 'a']++;
            countT[t.charAt(i) - 'a']++;
        }

        return Arrays.equals(countS, countT);
    }
}

// Test
class Test {
    public static void main(String[] args) {
        Solution solution = new Solution();

        System.out.println("Input: s = \"anagram\", t = \"nagaram\"");
        System.out.println("Output: " + solution.isAnagram("anagram", "nagaram"));
        System.out.println();

        System.out.println("Input: s = \"rat\", t = \"car\"");
        System.out.println("Output: " + solution.isAnagram("rat", "car"));
    }
}
